ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerVerticle.java

```diff
@@ -221,7 +196,46 @@ private static void unhandledExceptionHandler(final Throwable t) {
     log.error("Unhandled exception", t);
   }
 
-  private static Optional<AuthHandler> getAuthHandler(final Server server) {
+  private void setupAuthHandlers(final Router router) {
+    final Optional<Handler<RoutingContext>> jaasAuthHandler = getJaasAuthHandler(server);
+    final KsqlSecurityExtension securityExtension = server.getSecurityExtension();
+    final Optional<AuthenticationPlugin> authenticationPlugin = server.getAuthenticationPlugin();
+    final Optional<Handler<RoutingContext>> pluginHandler =
+        authenticationPlugin.map(plugin -> new AuthenticationPluginHandler(server, plugin));
+
+    if (jaasAuthHandler.isPresent() || authenticationPlugin.isPresent()) {
+      routeToNonProxiedEndpoints(router, ServerVerticle::pauseHandler);
+
+      routeToNonProxiedEndpoints(router,
+          rc -> wrappedAuthHandler(rc, jaasAuthHandler, pluginHandler));
+
+      // For authorization use auth provider configured via security extension (if any)
+      securityExtension.getAuthorizationProvider()
+          .ifPresent(ksqlAuthorizationProvider -> routeToNonProxiedEndpoints(router,
+              new KsqlAuthorizationProviderHandler(server.getWorkerExecutor(),
+                  ksqlAuthorizationProvider)));
+
+      routeToNonProxiedEndpoints(router, ServerVerticle::resumeHandler);
+    }
+  }
+
+  // If we have a Jaas handler configured and we have Basic credentials then we should auth
+  // with that
+  private static void wrappedAuthHandler(final RoutingContext routingContext,
+      final Optional<Handler<RoutingContext>> jaasAuthHandler,
+      final Optional<Handler<RoutingContext>> pluginHandler) {
+    if (jaasAuthHandler.isPresent()) {
+      final String authHeader = routingContext.request().getHeader("Authorization");
+      if (authHeader == null || authHeader.toLowerCase().startsWith("basic ")) {
+        jaasAuthHandler.get().handle(routingContext);
+        return;
+      }
+    }
+    pluginHandler
+        .ifPresent(routingContextHandler -> routingContextHandler.handle(routingContext));
+  }
```

Comment:
What happens if neither of these handlers is called (i.e., the jaasAuthHandler is present but an invalid auth header was provided, and no pluginHandler is present)? Do we appropriately fail the request in this case?

Reply:
Oops! Good catch